docs/source/tutorials/datasets.rst

```diff
@@ -108,33 +153,60 @@ Here is an example of sample that is formatted with :class:`~torchtune.data.Alpa
 
 We provide `other instruct templates <https://github.com/pytorch/torchtune/blob/main/torchtune/data/_instruct_templates.py>`_
 for common tasks such summarization and grammar correction. If you need to create your own
-instruct template for a custom task, you can create your own :class:`~torchtune.data.InstructTemplate`
-class and point to it in the config.
+instruct template for a custom task, you can inherit from :class:`~torchtune.data.InstructTemplate`
+and create your own class.
+
+.. code-block:: python
+
+    from torchtune.datasets import instruct_dataset
+    from torchtune.data import InstructTemplate
+
+    class CustomTemplate(InstructTemplate):
+        # Define the template as string with {} as placeholders for data columns
+        template = ...
+
+        # Implement this method
+        @classmethod
+        def format(
+            cls, sample: Mapping[str, Any], column_map: Optional[Dict[str, str]] = None
+        ) -> str:
+            ...
+
+    # Load in tokenizer
+    tokenizer = ...
+    dataset = instruct_dataset(
+        tokenizer=tokenizer,
+        source="my/dataset/path",
+        template="import.path.to.CustomTemplate",
```

Comment:
Why is this an import path and not an instance of CustomTemplate?

Reply:
because then I have to do a nested target in the config, which requires recursive instantiation. something we currently don't support.

In the `InstructDataset` class itself you pass in an instance of CustomTemplate, but the builders were made to be config friendly, hence the import path

Reply:
I don't know how I feel about this, it feels like we're saying we don't do recursion but then we do this slight of hand to get around it. The alternative would be that we instantiate the template in the recipe and pass it into the dataset.

Reply:
Yeah I'm not a fan of doing this, but instantiating it in the recipe means any user that is writing their own script will have to figure this out. Until we find an alternative to this sleight of hand, I would much rather handle it for the user. Happy to chat more about a more long-term solution